include/envoy/network/connection.h

```diff
@@ -139,6 +139,19 @@ class Connection : public Event::DeferredDeletable, public FilterManager {
    */
   virtual void noDelay(bool enable) PURE;
 
+#ifdef ENABLE_TCP_INFO
+  struct TCPInfo
+  {
+    uint32_t cwnd;      /* congestion window (packets) */
+    uint32_t in_flight; /* packets "in flight" */
+    uint32_t min_rtt;   /* minimum RTT in microsecond */
+    uint32_t rtt;       /* RTT in microsecond */
+    uint64_t delivery_rate;  /* bytes per second */
+  };
+
+  virtual absl::optional<TCPInfo> getTCPInfo() const PURE;
+#endif
```

Comment:
I like the idea of exposing this, but can we please do this in a dedicated PR? I would suggest defining a common TCPInfo struct as you have done, and then hiding it behind a platform API in which non-supported platforms can just return nullopt. We shouldn't expose the ifdefs to most of the code.

Reply:
Thanks for the comment Matt. You're right, we shouldn't expose ifdefs to most of the code. 
Let me work on it this week and submit another PR for TCPInfo :)